lib/optimize/ConcatenatedModule.js

```diff
@@ -0,0 +1,489 @@
+			info.source = resultSource;
+			info.globalScope = globalScope;
+			info.moduleScope = moduleScope;
+		});
+
+		// List of all used names to avoid conflicts
+		const allUsedNames = new Set(["__WEBPACK_MODULE_DEFAULT_EXPORT__", "defaultExport", "Object"]);
+
+		// get all global names
+		modulesWithInfo.forEach(info => {
+			info.globalScope.through.map(reference => reference.identifier.name).forEach(name => {
+				if(!/^__WEBPACK_MODULE_REFERENCE__\d+_(\d+|ns)__$/.test(name))
+					allUsedNames.add(name);
+			});
+		});
+
+		modulesWithInfo.forEach(info => {
+			const namespaceObjectName = this.findNewName("namespaceObject", allUsedNames, info.module.readableIdentifier(requestShortener));
+			info.internalNames.set(namespaceObjectName, namespaceObjectName);
+			info.exportMap.set(true, namespaceObjectName);
+			info.moduleScope.variables.forEach(variable => {
+				let name = variable.name;
+				if(allUsedNames.has(name)) {
+					const newName = this.findNewName(name, allUsedNames, info.module.readableIdentifier(requestShortener));
+					allUsedNames.add(newName);
+					info.internalNames.set(name, newName);
+					const source = info.source;
+					const allIdentifiers = new Set(variable.references.map(r => r.identifier).concat(variable.identifiers));
+					for(const identifier of allIdentifiers) {
+						const r = identifier.range;
+						source.replace(r[0], r[1] - 1, newName);
+					}
+				} else {
+					allUsedNames.add(name);
+					info.internalNames.set(name, name);
+				}
+			});
+		});
+
+		modulesWithInfo.forEach(info => {
+			info.globalScope.through.forEach(reference => {
+				const name = reference.identifier.name;
+				const match = /^__WEBPACK_MODULE_REFERENCE__(\d+)_(\d+|ns)__$/.exec(name);
+				if(match) {
+					const referencedModule = modulesWithInfo[+match[1]];
+					let exportName;
+					if(match[2] === "ns") {
+						exportName = true;
+					} else {
+						const exportIdx = +match[2];
+						exportName = referencedModule.module.providedExports[exportIdx];
+					}
+					const finalName = getFinalName(referencedModule, exportName, moduleToInfoMap, requestShortener);
+					if(!finalName)
+						throw new Error(`Cannot map to variable name in module ${info.module.resource} (export '${exportName}')`);
+					const r = reference.identifier.range;
+					const source = info.source;
+					source.replace(r[0], r[1] - 1, finalName);
+				}
+			});
+		});
+
+		const result = new ConcatSource();
+		if(moduleToInfoMap.get(this.rootModule).needCompatibilityFlag) {
+			result.add(`Object.defineProperty(${this.rootModule.moduleArgument || "exports"}, "__esModule", { value: true });\n`);
+		}
+		modulesWithInfo.forEach(info => {
+			result.add(`\n// CONCATENAMED MODULE: ${info.module.readableIdentifier(requestShortener)}\n`);
+			if(info.needNamespaceObject) {
+				const name = info.exportMap.get(true);
+				const nsObj = [`var ${name} = {};`];
```

Comment:
The exotic namespace object can be exactly reproduced, as far as it can be observed, by:

```js
const namespace = Object.create(null);
Object.defineProperty(namespace, Symbol.toStringTag, 'Module');
// add exports / getters
module.exports = Object.freeze(namespace);
```